### ISSUE TITLE
Storing secrets

### ISSUE THREAD

**Issue by tailhook** [tailhook/lithos#14]
# Motivation

In the past we have relied on keeping secrets in the filesystem and mounting them as:
```
volumes:
  /secrets: !Readonly /secrets
```

This is fine for smaller installations but has the following downsides:

1. Application must be able to read secrets from there
2. Users deploying applications are unaware of what is in the `/secrets` dir (i.e. is the config there is up to date)
3. They are plain-text for the host system
4. We need to sync them to machines somehow

# Solution

Add the following to `sandbox` config:
```
secrets-private-key: /etc/some/file
```
Add the following to `container` config:
```
secret-environ:
  SOME_SECRET: I7OO1RBTRYk+oAZ6n/dhRMCDXwgW
```
The value of `SOME_SECRET` is the actual value encoded by a **public key**.

When process is started, lithos will decode those values and pass them as environment variables to the process.

Upsides:
1. Passing secrets as env vars is known industry practice
2. It requires minimal support from applications (and most already support that)
3. Secrets are versioned even if the contents is unknown
4. Anybody with commit privileges can add a secret (no need to access server by SSH or other means)

Downsides:
1. You have to publish public key somehow
2. If private key is compromised you need to redeploy every project *(however, it's safe to assume that if private key is compromised, all the secrets are compromised, so you should change all the secrets anyway, presumably it's much more work than changing environment itself)*
3. Many keys will be duplicated if multiple lithos configs need same secrets, this is convenience vs explicitness trade-off

# Alternatives / Extra Features

We may also add the following to `sandbox`:
```
environ-secrets-file: /encrypted/secret.file # file is encrypted
```
And the following to `container` config:
```
external-secret-environ: [VAR1, VAR2]
```
To allow changing environ more dynamically.  But this means `/encrypted/secret.file` has to be deployed separately

Thoughts?

/cc @popravich, @anti-social, @jirutka, #13

**Comment by popravich** [comment]
In terms of the downsides (1) and (2) we can consider Travis' approach [described here](https://docs.travis-ci.com/user/encryption-keys/#Encryption-scheme), ie:
update our deployment tools to generate/update/revoke key-pair for roles being deployed.

**Comment by tailhook** [comment]
> In terms of the downsides (1) and (2) we can consider Travis' approach described here, ie:
update our deployment tools to generate/update/revoke key-pair for roles being deployed.

The problem we're facing is not encryption of secrets, but the distribution of secrets. If we're going to have a key per project we're going to face the same issue: how to distribute private keys.

Saying it in other words: every server having a project should have a private key for that project. And if we could securely and reliably deliver keys to that nodes, we could deliver secrets in the first place and not bother encryption (And reiterating on why security is basically equivalent: both ways are hackable by root user on a host system, and both cases are unhackable by other containers)

**Comment by tailhook** [comment]
This is done and released in lithos v0.16.0. Docs are [here](http://lithos.readthedocs.io/en/latest/tips/secrets.html)
